File: challenges/10/cpp/rs/main/bruijn.cpp
#include "bruijn.h"

ostream& operator<< (ostream& os, Bruijn& obj)
{
    obj.db(1,1);

    if(0 == obj.alphabet.size())
    {
        for(const auto &a : obj.sequence)
            os << '0'+a;
    }
    else
    {
        for(const auto &a : obj.sequence)
            os << obj.alphabet[a];
    }

    return os;
}

void Bruijn::db(unsigned int t, unsigned int p)
{
    if(t>n)
    {
        if(0==(n%p))
        {
            for(unsigned int i=1;i<p+1;++i)
                sequence.push_back(a[i]);
        }
    }
    else
    {
        a[t] = a[t - p];
        db(t + 1, p);

        for(unsigned int j = a[t - p] + 1; j < k ; ++j)
        {
            a[t] = j;
            db(t + 1, t);
        }
    }
}

File: challenges/10/cpp/rs/main/bruijn.py

def de_bruijn(k, n):
    """
    de Bruijn sequence for alphabet k
    and subsequences of length n.
    """
    try:
        # let's see if k can be cast to an integer;
        # if so, make our alphabet a list
        _ = int(k)
        alphabet = list(map(str, range(k)))

    except (ValueError, TypeError):
        alphabet = k
        k = len(k)

    a = [0] * k * n
    sequence = []

    def db(t, p):

        if t > n:
            if n % p == 0:
                sequence.extend(a[1:p + 1])
        else:
            a[t] = a[t - p]
            db(t + 1, p)
            for j in range(a[t - p] + 1, k):
                a[t] = j
                db(t + 1, t)
    db(1, 1)
    return "".join(alphabet[i] for i in sequence)

#print(de_bruijn("aBc", 6))
#print(de_bruijn("AaX", 3))
#print(de_bruijn(6, 12))

#print(de_bruijn(11, 2))

# För att testa seed ca 32 sec
#print(de_bruijn(5, 12))
print(de_bruijn("abcdef", 6))
